### ISSUE TITLE
LSP References of implementation are missed when stub files are used

### ISSUE THREAD

**Issue by gbelouze** [astral-sh/ty#4490]
### Summary

*This is an issue from an LSP perspective, though I suspect this is a duplicate for #1306 as a type checker perspective.*

When using stub files, LSP's find references from the `.py` implementation turn up empty.

## Minimal reproducer

```python
# a.pyi
class A: ...  # -> "find references" finds usage in b.py
```
```python
# a.py
class A:  # -> "find references" finds no references
    """Some documentation"""
    pass
```
```python
# b.py
from a import A
```

As far as I can tell, this is within the scope of expected usage for stub files. From the [typing specs](https://typing.python.org/en/latest/spec/distributing.html) : 

> [stub files] can provide type information for packages that do not wish to add them inline

 ( and later, as a recognized usage for type stubs)

> The package maintainer would like to add type information via stubs.

Note that *go to definition* (say, from `b.py`) still finds the `a.py` implementation, and *hover* still finds the documentation, so it's at least surprising that this doesn't hold for references.

## How other LSPs treat this

- `pyrefly` : same as `ty`
- `pyright` : same as `ty`
- `pylsp` : 
  - from `a.py`, only sees the `a.pyi` reference
  - from `a.pyi`, only sees the `a.py` reference (??)
  - from `b.py`, sees all occurrences of `A` in all three files

I'm surprised that none of those have the behavior that I would expect so maybe it's just me ?

(thank you for `ty`, it's a lovely tool ✨)
(Hopefully this is not me misconfiguring my LSPs and editor ✨)

### Version

ty 0.0.16 (e96ea4cd9 2026-02-10)

**Comment by MichaReiser** [comment]
Thanks for the detailed write up and comparing different LSPs. That's very helpful.

I agree, that this is very well in the realm of what you can expect from a type checker and we do something similar for go to declaration where we try to find the implementation for a symbol defined in a stub. 

The reason this doesn't work out of the box is that, from the type checkers perspective, `a.py` doesn't really matter because it uses `a.pyi` to resolve `A`. Therefore, nothing uses `a.py`, and find references returns no results. 

I'm sure there are some edge cases to this, but we could do the inverse of the goto-declaration stub mapping for find references (and rename). That is, if we're in a Python file, try to find the same symbol in a stub, and search for references to both definitions.
